From: Michal Hocko <mhocko@suse.com>
Date: Fri, 16 Jun 2017 00:06:28 +0200
Subject: fold me "mm: allow to configure stack gap size"
Bug-Debian-Security: https://security-tracker.debian.org/tracker/CVE-2017-1000364

- do not rely on is_stack when reporting the gap. show_map_vma has
  all the information we need
[carnil: backport for 3.16, adjust context, use of vm_is_stack in 3.16 rather
is_stack]
---
 fs/proc/task_mmu.c | 19 +++++++++++++------
 1 file changed, 13 insertions(+), 6 deletions(-)

--- a/fs/proc/task_mmu.c
+++ b/fs/proc/task_mmu.c
@@ -251,7 +251,7 @@ static int do_maps_open(struct inode *in
 }
 
 static void
-show_map_vma(struct seq_file *m, struct vm_area_struct *vma, int is_pid)
+show_map_vma(struct seq_file *m, struct vm_area_struct *vma, int is_pid, bool *has_gap)
 {
 	struct mm_struct *mm = vma->vm_mm;
 	struct file *file = vma->vm_file;
@@ -278,11 +278,17 @@ show_map_vma(struct seq_file *m, struct
 	start = vma->vm_start;
 	end = vma->vm_end;
 	if (vma->vm_flags & VM_GROWSDOWN) {
-		if (stack_guard_area(vma, start))
+		if (stack_guard_area(vma, start)) {
 			start = min(end, start + stack_guard_gap);
+			if (has_gap)
+				*has_gap = true;
+		}
 	} else if (vma->vm_flags & VM_GROWSUP) {
-		if (stack_guard_area(vma, end))
+		if (stack_guard_area(vma, end)) {
 			end = max(start, end - stack_guard_gap);
+			if (has_gap)
+				*has_gap = true;
+		}
 	}
 
 	seq_setwidth(m, 25 + sizeof(void *) * 6 - 1);
@@ -359,7 +365,7 @@ static int show_map(struct seq_file *m,
 	struct proc_maps_private *priv = m->private;
 	struct task_struct *task = priv->task;
 
-	show_map_vma(m, vma, is_pid);
+	show_map_vma(m, vma, is_pid, NULL);
 
 	if (m->count < m->size)  /* vma is copied successfully */
 		m->version = (vma != get_gate_vma(task->mm))
@@ -598,6 +604,7 @@ static int show_smap(struct seq_file *m,
 		.mm = vma->vm_mm,
 		.private = &mss,
 	};
+	bool has_gap = false;
 
 	memset(&mss, 0, sizeof mss);
 	mss.vma = vma;
@@ -605,7 +612,7 @@ static int show_smap(struct seq_file *m,
 	if (vma->vm_mm && !is_vm_hugetlb_page(vma))
 		walk_page_range(vma->vm_start, vma->vm_end, &smaps_walk);
 
-	show_map_vma(m, vma, is_pid);
+	show_map_vma(m, vma, is_pid, &has_gap);
 
 	seq_printf(m,
 		   "Size:           %8lu kB\n"
@@ -642,7 +649,7 @@ static int show_smap(struct seq_file *m,
 		seq_printf(m, "Nonlinear:      %8lu kB\n",
 				mss.nonlinear >> 10);
 
-	if (vm_is_stack(m->private, vma, is_pid))
+	if (has_gap)
 		seq_printf(m, "Stack_Gap:      %8lu kB\n", stack_guard_gap >>10);
 
 	show_smap_vma_flags(m, vma);
